### PR TITLE
many changes

### PR DESCRIPTION
**Pull request by magidnadav**
1) do not use classes for exceptions (use Error object instead)
2) more documentation
3) add support for intercepting async functions
4) handle the exception (do not support class exceptions)
5) add many tests

### CODE REVIEW

File: playback/tapeRecorder.ts [hunk 1]
```diff
@@ -1,54 +1,78 @@
...
+   */
+  enableRecording(): void {
     console.info("Enabling recording");
     this.recordingEnabled = true;
   }
 
-  play(recordingId: string, playbackFunction: Function): Playback {
-    const recording = this.tapeCassette.getRecording(recordingId);
+  /**
+   * Play again the recorder operation on current code
```
Comment (eitany) [inline playback/tapeRecorder.ts]: Plays the recorder operation using current code

File: playback/tapeRecorder.ts [hunk 2]
```diff
@@ -76,12 +105,15 @@ export class TapeRecorder {
       });
   }
 
+  /**
+   * Executes the operation function record its output and return the result
```
Comment (eitany) [inline playback/tapeRecorder.ts]: Executes the operation function, records

File: playback/tapeRecorder.ts [hunk 3]
```diff
@@ -91,26 +123,60 @@ export class TapeRecorder {
...
+      this.recordOutput({
+        alias: OPERATION_OUTPUT_ALIAS,
+        invocationNumber: 1,
+        args: [result],
+      });
+    };
+
+    let result: any;
+    try {
+      result = func(...args);
```
Comment (eitany) [inline playback/tapeRecorder.ts:145]: Did you check whether we should use func.apply here (and everywhere else) to make sure we are using the correct "this"
I have been listening in the fullstack interviews :)

File: playback/tapeRecorder.ts [hunk 4]
```diff
@@ -137,92 +208,263 @@ export class TapeRecorder {
...
+   */
   private playbackRecordedInterception(
     interceptionKey: string,
     args: any[]
   ): any {
     const recorded = this.playbackRecording!.getData(interceptionKey);
 
     if ("exception" in recorded) {
-      throw 
+      if (recorded.isError) {
```
Comment (eitany) [inline playback/tapeRecorder.ts:299]: Add a comment explaining why we need this special handle

File: playback/tapeRecorder.ts [hunk 5]
```diff
@@ -137,92 +208,263 @@ export class TapeRecorder {
...
+      //#endregion
+
+      if (that.inPlaybackMode()) {
+        try {
+          return that.playbackRecordedInterception(
+            interceptionKey as string,
+            args
+          );
+        } catch (error) {
+          if (error.name == PlaybackExceptionTypes.RecordingKeyError) {
```
Comment (eitany) [inline playback/tapeRecorder.ts]: This looked you ported the "run_intercepted_when_missing" part, but we didn't implement support for it and you implemented as if it defaults to True, while the default if False.
So you should remove this if part and raise error until we decide to implement this configurable capability

File: playback/tapeRecorder.ts [hunk 6]
```diff
@@ -137,92 +208,263 @@ export class TapeRecorder {
...
+      const interceptionKey =
+        that.outputInterceptionKey(params.alias, invocationNumber) + ".result";
 
       if (that.inPlaybackMode()) {
+        // Return recording of input invocation
         try {
-          return that.pla;
-        } catch (error) {}
+          return that.playbackRecordedInterception(interceptionKey, args);
+        } catch (error) {
```
Comment (eitany) [inline playback/tapeRecorder.ts]: Looks like redundant try catch (copy paste from port but we still didn't implement the catch configurable handling)

File: playback/tapeRecorder.ts [hunk 7]
```diff
@@ -137,92 +208,263 @@ export class TapeRecorder {
...
+
+  /**
+   * generate interception that do not do anything in playback mode.
+   * used for intercepting functions that should not be recorded .
+   * in playback mode, this function return playbackValue or an empty object
+   * @param func - the function to intercept
+   * @param playbackValue - the value to return in playback mode
+   * @returns the intercepted function
+   */
+  public muteInterception<inputType extends Array<any>, outputType>(
```
Comment (eitany) [inline playback/tapeRecorder.ts:324]: Please add tests for mute

File: playback/tapeRecorder.ts [hunk 8]
```diff
@@ -137,92 +208,263 @@ export class TapeRecorder {
...
+        );
       } catch (error) {
         const errorMessage = `Input interception key creation error for alias \'${
           params.alias
         }\' - ${JSON.stringify(error)}`;
 
         if (that.inPlaybackMode()) {
-          throw new InputInterceptionKeyCreationError(errorMessage);
+          throw generatePlaybackException(
+            PlaybackExceptionTypes.InputInterceptionKeyCreationError,
```
Comment (eitany) [inline playback/tapeRecorder.ts:374]: Let's add a test, you can use the key handler you added to test it

**Review by eitany — APPROVED** [review]

File: playback/tapeRecorder.ts (hunk 3 above)
Comment (magidnadav) [reply playback/tapeRecorder.ts:145]: after a talk with Daniel, we decided that the user needs to pass a bonded function (see example in the new tests I wrote) in order to make the code works.
the other option is to pass the function + the object and use apply which is almost the same but will make our code less readable

File: playback/tapeRecorder.ts [hunk 9]
```diff
@@ -91,26 +123,60 @@ export class TapeRecorder {
...
-    });
+    if (result && typeof result.then === "function") {
+      result.then(
+        (realResult: any) => {
+          handleResult(realResult);
+
+          return realResult;
+        },
+        (err: any) => {
+          handleException(err);
```
Comment (danielMiron) [inline playback/tapeRecorder.ts:158]: I think you are swallowing  the exception here 
original user will be unable to catch rejections 
`wrappedFn().then(...).catch((err)=> ...)`
catch will get the

**Review by danielMiron — APPROVED** [review]

**Pull request merged**